Add fats test for Reactve Messaging 3.0 Emitters

abutch3r opened the pull request:
Add some Lite mode tests for simple emitter checking with Kafka

Full Fat for Restful examples of using an emitter both positive and negative 

Fixes #25970

Azquelt committed:
Add support for nacks to the Kafka connector

Nacks are processed through a new MessageAccess utility which allows use
of the nack APIs while keeping the bundle compatible with the RM 1.0 API
so that the same code can be used for 1.0 and 3.0.

abutch3r committed:
Add Kafka Emitter test FATs

The tck does cover basically emitter functionality. but we need to test with the Liberty-kafka connector

Advanced scenarios for Nacking and using emitters in a REST resource

Reorg the kafka test package structure to simplify the structure

abutch3r committed:
Add Comments for how to handle Kafka resources

Update to resolve rebase issues

abutch3r commented:
#build

abutch3r commented:
#libby

tevans78 commented on dev/io.openliberty.microprofile.reactive.messaging.internal_fat/fat/src/io/openliberty/microprofile/reactive/messaging/fat/kafka/emitter/restful/KafkaEmitterRestfulTest.java:
```diff
@@ -0,0 +1,140 @@
+    public static LibertyServer server;
+
+    @ClassRule
+    public static final RepeatTests r = ReactiveMessagingActions.repeat(SERVER_NAME, ReactiveMessagingActions.MP61_RM30, ReactiveMessagingActions.MP50_RM30,ReactiveMessagingActions.MP60_RM30);
```
Do you need to run both MP60 and MP61 in every case? Maybe choose one or the other, it shouldn't make a huge difference for RM30.

posmith commented:
L2 message review completed.

abutch3r commented:
#libby 
#build

Azquelt commented on dev/io.openliberty.microprofile.reactive.messaging.internal_fat.common/src/com/ibm/ws/microprofile/reactive/messaging/fat/kafka/framework/KafkaTestClient.java:
```diff
@@ -72,6 +72,8 @@ public KafkaTestClient(String kafkaBootstrap) {
      * <p>
      * The connection properties are properties common to every consumer or producer connecting to this broker.
      *
+     * Once and instance of KafkaReader is no longer needed. It must be closed either via KafkaTestClient.cleanup() or KafkaReader.close()
+     *
```
Use `<p>` between paragraphs otherwise IDEs will lose the line breaks when displaying the doc for the method in a popup.

"Once and instance" -> "Once an instance"

Azquelt commented on dev/io.openliberty.microprofile.reactive.messaging.internal_fat/fat/src/io/openliberty/microprofile/reactive/messaging/fat/apps/emitter/EmitterRestResource.java:
```diff
@@ -0,0 +1,61 @@
+import java.util.concurrent.CompletionStage;
+
+@Path("/")
+public class EmitterRestResource<T> {
```
Not sure why `<T>` is used here?

Azquelt commented on dev/io.openliberty.microprofile.reactive.messaging.internal_fat/fat/src/io/openliberty/microprofile/reactive/messaging/fat/kafka/emitter/basic/KafkaEmitterTestServlet.java:
```diff
@@ -0,0 +1,66 @@
+
+    @Test
+    public void testmitterMessage() {
+        for(int i=1;i<=5;i++) {
```
Format file

Azquelt commented on dev/io.openliberty.microprofile.reactive.messaging.internal_fat/fat/src/io/openliberty/microprofile/reactive/messaging/fat/kafka/emitter/basic/KafkaEmitterTestServlet.java:
```diff
@@ -0,0 +1,66 @@
+    @Channel(CHANNEL_NAME2)
+    public Emitter<String> emitter2;
+
+    @Test
+    public void testmitterMessage() {
```
```suggestion
    @Test
    public void testEmitterMessage() {
```

Azquelt commented on dev/io.openliberty.microprofile.reactive.messaging.internal_fat/fat/src/io/openliberty/microprofile/reactive/messaging/fat/kafka/emitter/nack/KafkaEmitterNackRestfulTest.java:
```diff
@@ -0,0 +1,129 @@
+    public void testEmittingNackPayload() throws Exception {
+        server.setMarkToEndOfLog();
+        int port = server.getHttpDefaultPort();
+        URL url = new URL("http://localhost:"+port+"/"+APP_NAME+"/payload");
+        HttpURLConnection conn = HttpUtils.getHttpConnection(url, HttpUtils.DEFAULT_TIMEOUT, HttpUtils.HTTPRequestMethod.POST);
+        OutputStream os = conn.getOutputStream();
+        os.write(("test").getBytes());
+        os.flush();
+        assertThat(conn.getResponseCode(), is(500));
+        os.close();
+        conn.disconnect();
+        // check that we get error about the payload emitter channel
+        assertNotNull("Channel can ",server.waitForStringInLogUsingMark("CWMRX1003E.*restful-emitter-payload"));
```
This is a bit of a faff isn't it.

Tempted to suggest we should try-catch around the IO to ensure streams are closed, but in a test like this it's _probably_ not a big problem.

If we end up doing this a lot, I might be tempted to extend `componenttest.topology.utils.HttpRequest` to add more general support for POST requests, but otherwise it can be left as-is.

Azquelt replied:
~Another option would be to change the rest resource to not use POST and then use `HttpRequest` as-is.~

I see in the next test why you need to pass in a body.

Azquelt commented on dev/io.openliberty.microprofile.reactive.messaging.internal_fat/fat/src/io/openliberty/microprofile/reactive/messaging/fat/kafka/emitter/nack/KafkaEmitterNackRestfulTest.java:
```diff
@@ -0,0 +1,129 @@
+    public void testEmittingNackPayload() throws Exception {
+        server.setMarkToEndOfLog();
+        int port = server.getHttpDefaultPort();
+        URL url = new URL("http://localhost:"+port+"/"+APP_NAME+"/payload");
```
I suspect this will always work, but just use `HttpUtils.createURL` to build the URL rather than substituting in the port yourself.

Azquelt commented on dev/io.openliberty.microprofile.reactive.messaging.internal_fat/fat/src/io/openliberty/microprofile/reactive/messaging/fat/kafka/emitter/nack/KafkaEmitterNackRestfulTest.java:
```diff
@@ -0,0 +1,129 @@
+
+    @AfterClass
+    public static void teardown() throws Exception {
+        try {
+        //We expect a kafka error and an error from RESTEASY relating to th failure when we respond to the request
+        server.stopServer("CWMRX1003E.*restful-emitter", "RESTEASY002020");
+        } finally {
```
indentation

Azquelt commented on dev/io.openliberty.microprofile.reactive.messaging.internal_fat/fat/src/io/openliberty/microprofile/reactive/messaging/fat/kafka/emitter/restful/KafkaEmitterRestfulTest.java:
```diff
@@ -0,0 +1,140 @@
+        int port = server.getHttpDefaultPort();
+        URL url = new URL("http://localhost:"+port+"/"+APP_NAME+"/"+path);
+        // Send 5 messages to the server to make sure we do process multiple requests
+        for(int count = 1; count <6;count++) {
```
Format file

Azquelt commented on dev/io.openliberty.microprofile.reactive.messaging.internal_fat/fat/src/io/openliberty/microprofile/reactive/messaging/fat/kafka/emitter/restful/KafkaEmitterRestfulTest.java:
```diff
@@ -0,0 +1,140 @@
+    }
+
+
+    public void sendRequest(String path) throws IOException {
+        int port = server.getHttpDefaultPort();
+        URL url = new URL("http://localhost:"+port+"/"+APP_NAME+"/"+path);
```
Same comments about building the URL apply here too.

Azquelt commented on dev/io.openliberty.microprofile.reactive.messaging.internal_fat/fat/src/io/openliberty/microprofile/reactive/messaging/fat/kafka/emitter/restful/KafkaEmitterRestfulTest.java:
```diff
@@ -0,0 +1,140 @@
+    }
+
+
+    public void sendRequest(String path) throws IOException {
```
```suggestion
    public void sendRequests(String path) throws IOException {
```

Azquelt commented on dev/io.openliberty.microprofile.reactive.messaging.internal_fat/fat/src/io/openliberty/microprofile/reactive/messaging/fat/kafka/nack/KafkaNackTest.java:
```diff
@@ -97,10 +98,11 @@ public static void setup() throws Exception {
     public static void teardown() throws Exception {
         server.stopServer("CWMRX1003E.*nack-test-channel", // CWMRX1003E: An error occurred when sending a message to the Kafka broker. The error is: org.apache.kafka.common.errors.TimeoutException: Topic nack-test-channel not present in metadata after 5000 ms.
                           "CWMRX1011E.*KafkaNackTestException");
+        KafkaUtils.deleteKafkaTopics(KafkaTests.getAdminClient());
     }
```
Do we usually do the deletion in a try-finally around `stopServer()`?

It's common for `stopServer` to throw an exception if the test fails or some infrastructure breaks which might cause knock-on failures here?

Azquelt commented on dev/io.openliberty.microprofile.reactive.messaging.internal_fat/publish/servers/SimpleRxMessagingServer/server.xml:
```diff
@@ -16,9 +16,11 @@
         <feature>componenttest-2.0</feature>
-        <feature>osgiconsole-1.0</feature>
         <feature>localConnector-1.0</feature>
+        <feature>mpMetrics-4.0</feature>
         <feature>mpReactiveMessaging-3.0</feature>
+        <feature>osgiconsole-1.0</feature>
+        <feature>restfulWS-3.1</feature>
         <feature>servlet-5.0</feature>
         <feature>mpConfig-3.1</feature>
         <feature>mpMetrics-5.0</feature>
```
You have `mpMetrics-4.0` and `mpMetrics-5.0` in the server.xml.

Azquelt commented on dev/io.openliberty.microprofile.reactive.messaging.internal_fat/fat/src/io/openliberty/microprofile/reactive/messaging/fat/apps/emitter/EmitterRestResource.java:
```diff
@@ -0,0 +1,61 @@
+    public static final String PAYLOAD_CHANNEL_NAME = "restful-emitter-payload";
+    public static final String MESSAGE_CHANNEL_NAME = "restful-emitter-message";
+
+    @Inject
+    @Channel(PAYLOAD_CHANNEL_NAME)
+    Emitter<String> payloadEmitter;
+
+    @Inject
+    @Channel(MESSAGE_CHANNEL_NAME)
+    Emitter<String> messageEmitter;
```
Use `private` on injected fields unless you're deliberately allowing them to be accessed from elsewhere.

Azquelt reviewed (changes requested):
Tests look good.

Comments are just cleanup and nitpicks.

abutch3r committed:
Address comments

- Remove double mpMetrics definitions in server.xml
- tidy up comments on common classes
- resolve exposure of Emitter resources
- Improve code resilience

abutch3r commented:
#libby
#build
